devices/wholeBodyDynamics/WholeBodyDynamicsDevice.cpp

```diff
@@ -303,78 +303,257 @@ bool WholeBodyDynamicsDevice::openEstimator(os::Searchable& config)
+        nrUnknownsInExtWrench.erase(nrUnknownsInExtWrench.begin() + indexesToDiscard[i]);
+    }
+
+    // The frames from in 1 submodel should be no more than 1 if the type is FULL_WRENCH, no more than 2 if the type is PURE_FORCE, and no more than 6 if the type is PURE_FORCE_WITH_KNOWN_DIRECTION.
+    // Refer to section 3 of the theory paper https://www.researchgate.net/publication/236152161_Contact_Force_Estimations_Using_Tactile_Sensors_and_Force_Torque_Sensors
+
     // For each submodel, we find the first suitable contact frame
     // This is a n^2 algorithm, but given that is just used in
     // configuration it should be ok
     size_t nrOfSubModels = estimator.submodels().getNrOfSubModels();
 
-    // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a default contact for the given submodel
-    subModelIndex2DefaultContact.resize(nrOfSubModels,iDynTree::FRAME_INVALID_INDEX);
+    // We indicate with FRAME_INVALID_INDEX the fact that we still don't have a default/override contact for the given submodel
+    contactFramesIdxValidForSubModel.resize(contactFramesIdx.size(),iDynTree::FRAME_INVALID_INDEX);
+    //create a vector to store the submodel indexes that will have contact points
+    std::vector<size_t> validSubModelsIndexes;
+    validSubModelsIndexes.resize(contactFramesIdx.size(),iDynTree::FRAME_INVALID_INDEX);
+
+    nrUnknownsInSubModel.resize(nrOfSubModels,0);
 
-    for(size_t i=0; i < defaultContactFramesIdx.size(); i++)
+    for(size_t i=0; i < contactFramesIdx.size(); i++)
     {
-        size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),defaultContactFramesIdx[i]);
+        size_t subModelIdx = estimator.submodels().getSubModelOfFrame(estimator.model(),contactFramesIdx[i]);
 
-        // If the subModel of the frame still does not have a default contact, we add it
-        if( subModelIndex2DefaultContact[subModelIdx] == iDynTree::FRAME_INVALID_INDEX )
+        //check if adding the current contact to this subModel will increase the no of variable to more than 6
+        if(nrUnknownsInSubModel[subModelIdx]+nrUnknownsInExtWrench[i] > 6)
         {
-            subModelIndex2DefaultContact[subModelIdx] = defaultContactFramesIdx[i];
+            yWarning() << "subModel no " << subModelIdx << " has the maximum number of variables (6). The frame " << contactFramesNames[i] << " will not be added to it.";
+        }
+        else
+        {
+            //add the contact frame
+            contactFramesIdxValidForSubModel[i] = contactFramesIdx[i];
+            //add the submodel index to the valid submodel indexes group
+            validSubModelsIndexes[i] = subModelIdx;
+            nrUnknownsInSubModel[subModelIdx] += nrUnknownsInExtWrench[i];
+            yInfo() << "Adding frame " << contactFramesNames[i] << " (idx= " << contactFramesIdx[i] << ") for subModel no " << subModelIdx << " Number of unknowns has become= " << nrUnknownsInSubModel[subModelIdx];
```

Comment:
I guess this is a debug output that can be removed?

Reply:
I actually found it useful also as a user. It gives you a quick check of the contacts added instead of going through the configuration file to do proof reading.

Reply:
I agree, but then we should print that for all the configuration files.

Reply:
Okay to keep the consistency I removed the line in https://github.com/robotology/whole-body-estimators/pull/48/commits/a91ed5422a088256a65c7c27d838184f94cb28ba